applications: nrf_desktop: Power Manager documentation (#2102)
NordicPlayground/fw-nrfconnect-nrf: applications/nrf_desktop/doc/power_manager.rst

.. _nrf_desktop_power_manager:

Power manager module
####################

> Use the ``power_manager`` module to reduce the energy consumption of battery-powered devices.

Reviewer: ```suggestion
Use the power manager module to reduce the energy consumption of battery-powered devices.
```
Use the module name, not its code name in the docs (except when discussing code).

This is achieved by switching to low power modes when the device is not used for a longer period of time.

> 

Reviewer: Here it's OK not to use the empty line: "this" rephrases the content of the previous sentence, and new paragraph would make this syntactically weird.

> 

Author: The idea was to keep it in the same paragraph.

Module Events
*************

+----------------------------------+------------------------+----------------------------------+----------------------+------------------------------------+
| Source Module                    | Input Event            | This Module                      | Output Event         | Sink Module                        |
+==================================+========================+==================================+======================+====================================+

> |                                  |                        | :ref:`nrf_desktop_power_manager` | ``power_down_event`` | :ref:`nrf_desktop_power_manager`   |

Reviewer 2: Some interactions are missing here.

Author: yes - you are generating the table so its pointless to fill everything, right ;)

+----------------------------------+------------------------+                                  +----------------------+------------------------------------+
| :ref:`hid_state`                 | ``hid_report_event``   |                                  |                      |                                    |
+----------------------------------+------------------------+                                  +----------------------+------------------------------------+
| :ref:`nrf_desktop_power_manager` | ``power_down_event``   |                                  |                      |                                    |
+----------------------------------+------------------------+                                  +----------------------+------------------------------------+
|                                  | ``wake_up_event``      |                                  |                      |                                    |
+----------------------------------+------------------------+----------------------------------+----------------------+------------------------------------+

> External dependencies

Reviewer: I see this section only in this module?

I would move this whole section to Implementation details or Configuration. After all, these are details about how this module is implemented in nRF Desktop, and partially introduce what you need to select (CONFIG_DESKTOP_POWER_MANAGER_ENABLE) and configure...

Author: This is copy paste from watchdog which I used as a base. Will move it and do the same for the watchdog.

Author: Done

*********************

This module utilizes Zephyr :ref:`zephyr:power_management`.

> It depends on :option:`HAS_SYS_POWER_STATE_DEEP_SLEEP_1` being enabled and selects :option:`DEVICE_POWER_MANAGEMENT` and :option:`SYS_POWER_DEEP_SLEEP_STATES`.

Reviewer 2: Kconfig options should have `CONFIG_` prefix

Reviewer 2: Maybe we could merge `External dependencies` with the `Configuration` option. To make it consistent with other modules.

Author: Fixed

Configuration
*************

> The module is enabled by the ``CONFIG_DESKTOP_POWER_MANAGER_ENABLE`` option.

Reviewer: Is this option something the user needs to enable or it is enabled by default? Current wording can suggest the enabled-by-default scenario.

Author: Reworded

With ``CONFIG_DESKTOP_POWER_MANAGER_TIMEOUT`` configuration option you can set time after which the application will go to low power mode.

> The ``CONFIG_DESKTOP_POWER_MANAGER_STAY_ON`` let system stay idle also when there are no active connections.

Reviewer: ```suggestion
The ``CONFIG_DESKTOP_POWER_MANAGER_STAY_ON`` lets the system stay idle also when there are no active connections.
```

> The ``CONFIG_DESKTOP_POWER_MANAGER_ERROR_TIMEOUT`` set the time after which the device is turned off upon an internal error.

Reviewer: Are these three lines "suboptions" of the enable option? If so, I would introduce them with a sentence ("Once the module is enabled, you can enable one or more of the following options:") and put these three into bullet points.

> The ``CONFIG_DESKTOP_POWER_MANAGER_ERROR_TIMEOUT`` set the time after which the device is turned off upon an internal error.

Reviewer: ```suggestion
The ``CONFIG_DESKTOP_POWER_MANAGER_ERROR_TIMEOUT`` sets the time after which the device is turned off upon an internal error.
```

Implementation details
**********************

> The :ref:`nrf_desktop_power_manager` is started when the :ref:`nrf_desktop_main` module is ready (reported using ``module_state_event``).

Reviewer: Do not link a page to itself, please.

Author: Fixed

> The :ref:`nrf_desktop_power_manager` takes control of the overall system power state.

Reviewer 2: I think it may be good to introduce all the power states here and refer to them later in this page - for example there is no information about the idle state (its not system power state, but it affects the device a lot).
Otherwise this could be confusing to the user.

Author: This can be useful.

Author: I don't think I would give every internal state but a list of states can be easier to perceive compared to block of text.

Author: Re-written this part. It should be easier to read now.

> Except when entering a deep sleep mode the system will stay in ``SYS_POWER_STATE_ACTIVE`` state.

Reviewer: ```suggestion
The system will always stay in the ``SYS_POWER_STATE_ACTIVE`` state.
The only exception is when it enters the deep sleep mode.
```

> When entering the deep sleep mode the module will set system to ``SYS_POWER_STATE_DEEP_SLEEP_1`` state.

Reviewer: ```suggestion
In such case, the module will set the system to the ``SYS_POWER_STATE_DEEP_SLEEP_1`` state.
```

> Power down

Reviewer: I feel a bit confused, because to power down means to turn off, but switching to low power is not exactly turning off? I suggest you rephrase the section title.

Author: I have rephrased this part.

Reviewer: Would it be ok to change this section name to "Switching to low power" or "Low power scenarios"?

Author: I think so

Author: Done

==========

> If the device is not connected to the USB, the module counts time since the last user interaction (last HID report sent from the device).

Reviewer: Can you add an introductory sentence that states why the power down is dependant on the USB connection, and also introduces the two options below?

Reviewer: Also, can you add an introductory sentence about what is discussed in this section?

Maybe by also redo the whole section into unordered steps or provide an overview at the start of what is going on during power down:
- checking USB
- setting the low power mode
- checking BLE
- reporting low power

> If the device is not connected to the USB, the module counts time since the last user interaction (last HID report sent from the device).

Reviewer: ```suggestion
If the device is not connected to a USB, the module counts time since the last user interaction (that is, since the last HID report sent from the device).
```

If the device is connected to the USB, the module remains idle.

> On timeout :ref:`nrf_desktop_power_manager` sets the application to the low power state.

Reviewer: What is the relation of this paragraph to the previous one? Is the timeout referring to counting the time from the last interaction?

> On timeout :ref:`nrf_desktop_power_manager` sets the application to the low power state.

Reviewer: ```suggestion
On timeout, the power manager module sets the application to the low power state.
```

> When switching to the low power state :ref:`nrf_desktop_power_manager` will send a `power_down_event``.

Reviewer: ```suggestion
When switching to the low power state, the module will send a `power_down_event``.
```

> Other application modules react to this event to change their configuration to low power (e.g. turn off LEDs).

Reviewer: ```suggestion
Other application modules react to this event by changing their configuration to low power, for example by turning off LEDs.
```

> Only after all modules confirmed entering the low power state (by not consuming the ``power_down_event``) the :ref:`nrf_desktop_power_manager` will sets application state to low power.

Reviewer: ```suggestion
Only after all modules confirmed that they have entered the low power state (by not consuming the ``power_down_event``), the power manager will set the application's state to low power.
```

> The next step depends on the BLE connection state.

Reviewer: Add empty line after this line. Otherwise the list formatting breaks. See templates/cheat sheet for info.

>  * If the device is connected to the BLE, the module will perform no more actions and the system will remain in an idle state.

Reviewer: No spaces before * in lists in .rst.

>  * If the device is connected to the BLE, the module will perform no more actions and the system will remain in an idle state.

Reviewer: ```suggestion
* If the device is connected to the BLE, the module will not perform other actions and the system will remain in the idle state.
```

>  * If the device is not connected, the module will switch the system into the deep sleep mode.

Reviewer: ```suggestion
* If the device is not connected, the module will switch the system to the deep sleep mode.
```

>    When entering the deep sleep mode, Zephyr suspends all devices and CPU is set to OFF state.

Reviewer: ```suggestion
   When entering the deep sleep mode, Zephyr suspends all devices and CPU is set to the OFF state.
```

>    To exit deep sleep mode CPU needs an external event.

Reviewer: ```suggestion
   To exit the deep sleep mode, the CPU needs an external event.
```

> If disconnection happens while device is in the low power mode, the :ref:`nrf_desktop_power_manager` will set system to the deep sleep mode as described above.

Reviewer: ```suggestion
If a disconnection happens while the device is in the low power mode, the power manager will set the system to the deep sleep mode as described above.
```

> The application can be configured to keep system in the idle state, instead of entering the deep sleep mode, when there are no active connections.

Reviewer: ```suggestion
The application can be configured to keep the system in the idle state when there are no active connections, instead of entering the deep sleep mode.
```

> To select this behavior use ``CONFIG_DESKTOP_POWER_MANAGER_STAY_ON`` configuration option.

Reviewer: ```suggestion
To select this behavior, use the ``CONFIG_DESKTOP_POWER_MANAGER_STAY_ON`` configuration option.
```

Each module must report entering the low power state by sending the appropriate module state event, when it has done so.

Postponed power down
--------------------

> When :ref:`nrf_desktop_power_manager` detects that the application should enter the low power state it sends the ``power_down_event``.

Reviewer: ```suggestion
When the power manager detects that the application is to enter the low power state, it sends the ``power_down_event``.
```

> It is possible that some module will not be ready to switch to the lower power state at that time.

Reviewer: ```suggestion
It is possible that some modules will not be ready to switch to the lower power state at that time.
```

> In such case this module should consume the ``power_down_event`` and change its internal state so it enters the low power state when ready.

Reviewer: ```suggestion
In such case, the module that is not yet ready should consume ``power_down_event`` and change its internal state, so that it enters the low power state when ready.
```
Is there a less figurative way of saying "to consume event"? "complete"? "accept"?

> 
> The :ref:`nrf_desktop_power_manager` will continue with the low power state change, when it gets notification that any module switched for the low power.

Reviewer: ```suggestion
The power manager will continue with the low power state change when it gets a notification that the module switched to the low power.
```

> Wake up

Reviewer: ```suggestion
Wake-up scenarios
```

=======

> To exit low power mode the application waits for a wake up event.

Reviewer: ```suggestion
To exit the low power mode, the application waits for a wake-up event.
```

> The other application modules, upon entering the low power state, can configure hardware to make sure such event is received.

Reviewer 2: "to generate/submit such event"

Author: Can you elaborate?

Reviewer: Do not use slash. Use `X or Y, or both" if you want to say "and/if".

Reviewer 2: I was thinking that it may be better to write here something like:

The other application modules, upon entering the low power state, can configure hardware to generate such event.

To inform that the event is simply generated by one of our modules.

> When system is in the deep sleep mode, wake up interrupt will reboot the system.

Reviewer: ```suggestion

When the system is in the deep sleep mode, a wake-up interrupt will reboot the system.
```

> After the reboot the application initializes itself again along with the entire Zephyr.

Reviewer: ```suggestion
After the reboot, the application initializes itself again, along with the Zephyr modules.
```

> When system was in the idle state, the other module must submit a ``wake_up_event`` upon reception of hardware interrupt.

Reviewer: ```suggestion
If the system was in the idle state, the other module must submit a ``wake_up_event`` upon the reception of hardware interrupt.
```

> The ``wake_up_event`` is received by the application modules, switching them back to the normal operation.

Reviewer: ```suggestion
The ``wake_up_event`` is received by the application modules and it switches them back to the normal operation.
```

> The :ref:`nrf_desktop_power_manager` will set application state to active and re-start low power timeout counter (if not connected to the USB).

Reviewer: ```suggestion
The power manager will set the application state to active and restart the low power timeout counter (if not connected to the USB).
```
What if connected to the USB? Don't leave this implicit, please.

Error handling
==============

> The :ref:`nrf_desktop_power_manager` checks if any application modules reported an error condition.

Reviewer: ```suggestion
The power manager module checks if any application modules reported an error condition.
```
When does it do so? And why? Not clear.

> When any module switches to the error state, :ref:`nrf_desktop_power_manager` will put the system to the deep sleep after time defined by ``CONFIG_DESKTOP_POWER_MANAGER_ERROR_TIMEOUT``.

Reviewer: ```suggestion
When a module switches to the error state, the power manager will put the system to the deep sleep after the amount of time defined by ``CONFIG_DESKTOP_POWER_MANAGER_ERROR_TIMEOUT``.
```

> During this time error condition can be reported to the used by other modules (e.g. :ref:`led_state`).

Reviewer: ```suggestion
During this period, the error condition can be reported to the used by other modules (for example, :ref:`led_state`).
```
`to the used` -- looks like a word is missing here?

+----------------------------------+------------------------+                                  +----------------------+------------------------------------+
| ``hid_state``                    | ``hid_report_event``   |                                  |                      |                                    |
+----------------------------------+------------------------+                                  +----------------------+------------------------------------+
| ``power_manager``                | ``power_down_event``   |                                  |                      |                                    |
+----------------------------------+------------------------+                                  +----------------------+------------------------------------+
|                                  | ``wake_up_event``      |                                  |                      |                                    |
+----------------------------------+------------------------+----------------------------------+----------------------+------------------------------------+

Configuration
*************

This module uses Zephyr's :ref:`zephyr:power_management` subsystem.

> It depends on :option:`CONFIG_HAS_SYS_POWER_STATE_DEEP_SLEEP_1` being enabled and selects :option:`CONFIG_DEVICE_POWER_MANAGEMENT` and :option:`CONFIG_SYS_POWER_DEEP_SLEEP_STATES`.

Reviewer 2: The dependencies are related to `CONFIG_DESKTOP_POWER_MANAGER_ENABLE`. We should move the next line here I think.

Author: I have put it at the first line now.

You can enable the power manager module by selecting the ``CONFIG_DESKTOP_POWER_MANAGER_ENABLE`` option in the configuration.

Timeout configuration options
    With the ``CONFIG_DESKTOP_POWER_MANAGER_TIMEOUT`` configuration option you can set time after which the application will go to low power mode.
    By default, the timeout is set to 120 seconds.

    The ``CONFIG_DESKTOP_POWER_MANAGER_ERROR_TIMEOUT`` sets the time after which the device is turned off upon an internal error.

Optional idle boolean

>     The ``CONFIG_DESKTOP_POWER_MANAGER_STAY_ON`` lets the system stay idle also when there are no active connections.

Author: I have changed `idle` to `on` to avoid confusion. I don't want to use names that were yet not introduced. The effect of this option is clearly explained later on.

For more information about configuration options, check help in the configuration tool.

Implementation details
**********************

The power manager is started when the :ref:`nrf_desktop_main` module is ready (reported using ``module_state_event``).

When started, it can do the following operations:

* Manage `Power states`_
* Trigger `Power down`_
* Handle `Wake-up scenarios`_ from suspended or off states

Power states
============

The application can be in the following power states:

 * `Idle`_
 * `Suspended`_
 * `Off`_
 * `Error`_

The power manager takes control of the overall operating system power state.
See the following section for more details about how the application states converts to the system power state.

Idle
----

In this state, the board is either connected or actively looking for a new connection (advertising).
The peripherals can be turned on, including LEDs.

The application remains in this state indefinitely if the device is connected to the USB.

> In such case, the operating system will be kept in the ``SYS_POWER_STATE_ACTIVE`` state.

Reviewer: Is this correct, @pdunaj ? The state is `Idle`, but the state define is `SYS_POWER_STATE_ACTIVE`?

Author: yes, idle is the application state while `SYS_POWER_STATE_ACTIVE` is operating system state

If the device is not connected to a USB, the module counts time elapsed since the last user interaction (that is, since the last HID report sent from the device).
On timeout, the power manager module sets the application to either the suspended or the off state.

Suspended
---------

> Upon power-down timeout, the power manager will switch the application to the low power state if the following conditions are met:

Author: Fixed

> * The device is connected.

Author: changed, but only in this place

* The option ``CONFIG_DESKTOP_POWER_MANAGER_STAY_ON`` is selected.

The other modules of the application, if applicable, will turn off the peripherals or switch them to standby to conserve power.
The operating system will be kept in the ``SYS_POWER_STATE_ACTIVE`` state.

It is assumed that the operating system will conserve power by setting the CPU state to idle whenever possible.
The established connection is maintained.

Off
---

Upon power-down timeout, the power manager will switch the application to the deep sleep mode if the following conditions are met:

* The device is disconnected.
* The option ``CONFIG_DESKTOP_POWER_MANAGER_STAY_ON`` is disabled.

The other modules of the application, if applicable, will turn off the peripherals or switch them to standby to conserve power.
The operating system will be switched to the ``SYS_POWER_STATE_DEEP_SLEEP_1`` state.
The devices will be suspended and CPU will be switched to the deep sleep (off) mode.

A device reboot is requited to exit this state.

Error
-----

> In this state, the power manager module checks if any application modules reported an error condition.

Author: True. This state is set by the power manager in cases of error in any of the modules. Maybe I have messed up something in the original text.

> In this state, the power manager module checks if any application modules reported an error condition.

Author: Reworded.

> When a module switches to the error state, the power manager will put the application to off state after the amount of time defined by ``CONFIG_DESKTOP_POWER_MANAGER_ERROR_TIMEOUT``.

Reviewer: Ok, I might've got it wrong when editing, or something is not clear here: line 136 says "In this state, the power manager", which suggests it's a state of the power manager.

Then, in line 138, we have "When a module switches to the error state". So I'm confused whether this is the description of the manager's Error state or modules' Error state.

Author: I put `a module`, meaning any module in the application. Error state in this context is the internal state of the module passed with `module_state_event`

Author: I have made it more explicit now.

In such case, the module that is not yet ready should consume ``power_down_event`` and change its internal state, so that it enters the low power state when ready.

Each module must report entering the low power state by sending ``module_state_event``, when it has done so.
The power manager will continue with the low power state change when it gets a notification that the module switched to the low power.

Only after all modules confirmed that they have entered the low power state (by not consuming the ``power_down_event``), the power manager will set the required application's state.

If a disconnection happens while the device is in the suspended state, the power manager will switch the application to the off state.
The application can be configured to keep the system in the suspended state when there are no active connections instead.
To select this behavior, use the ``CONFIG_DESKTOP_POWER_MANAGER_STAY_ON`` configuration option.

Wake-up scenarios
=================

The application can be woken up in the following scenarios:

* `Wake-up from suspended state`_
* `Wake-up from off state`_

Wake-up from suspended state
----------------------------

> Any module can trigger the application to switch from the suspended to the idle state by submitting a ``wake_up_event``.

Reviewer: ```suggestion
Any module can trigger the application to switch from the suspended state back to the idle state by submitting a ``wake_up_event``.
```

When started, it can do the following operations:

* Manage `Power states`_
* Trigger `Power down`_
* Handle `Wake-up scenarios`_ from suspended or off states

Power states
============

The application can be in the following power states:

 * `Idle`_
 * `Suspended`_
 * `Off`_
 * `Error`_

The power manager takes control of the overall operating system power state.
See the following section for more details about how the application states converts to the system power state.

Idle
----

In this state, the board is either connected or actively looking for a new connection (advertising).
The peripherals can be turned on, including LEDs.

The application remains in this state indefinitely if the device is connected to the USB.
In such case, the operating system will be kept in the ``SYS_POWER_STATE_ACTIVE`` state.

If the device is not connected to a USB, the module counts time elapsed since the last user interaction (that is, since the last HID report sent from the device).
On timeout, the power manager module sets the application to either the suspended or the off state.

Suspended
---------

> Upon power-down timeout, the power manager will switch the application to the low power state if one of the following conditions are met:

Author: done

* The device is connected to a remote peer.
* The option ``CONFIG_DESKTOP_POWER_MANAGER_STAY_ON`` is selected.

The other modules of the application, if applicable, will turn off the peripherals or switch them to standby to conserve power.
The operating system will be kept in the ``SYS_POWER_STATE_ACTIVE`` state.

It is assumed that the operating system will conserve power by setting the CPU state to idle whenever possible.
The established connection is maintained.

Off
---

Upon power-down timeout, the power manager will switch the application to the deep sleep mode if the following conditions are met:

* The device is disconnected.
* The option ``CONFIG_DESKTOP_POWER_MANAGER_STAY_ON`` is disabled.

The other modules of the application, if applicable, will turn off the peripherals or switch them to standby to conserve power.
The operating system will be switched to the ``SYS_POWER_STATE_DEEP_SLEEP_1`` state.
The devices will be suspended and CPU will be switched to the deep sleep (off) mode.

> A device reboot is requited to exit this state.

Author: done

Error
-----

The power manager module checks if any application modules reported an error condition.

When any application module switches to the error state (broadcasts ``MODULE_STATE_ERROR`` via ``module_state_event``), the power manager will put the application to error state.
Then, after the amount of time defined by ``CONFIG_DESKTOP_POWER_MANAGER_ERROR_TIMEOUT``, it will put the application to off state.

> During this period, the error condition can be reported by other modules (for example, :ref:`nrf_desktop_led_state`).

Reviewer 2: `displayed to the user`

Author: Done
